### PR TITLE
[traffic-gen-in-vm] Run traffic towards the VMI under test

### PR DESCRIPTION
**Pull request by RamLavi**
This PR is adding the command that will initiate the DPDK traffic towards the VMI under test, and in the meanwhile will poll the traffic for drop rates on the traffic-generator side.

### CODE REVIEW

File: pkg/internal/checkup/executor/executor.go [hunk 1]
```diff
@@ -109,6 +109,12 @@ func (e Executor) Execute(ctx context.Context, vmiUnderTestName, trafficGenVMINa
 		return status.Results{}, err
 	}
 
+	log.Printf("Clearing Trex console stats before test...")
```
Comment (orelmisan) [inline pkg/internal/checkup/executor/executor.go:107]: nit: Commit message:
> trex, executor: Clear stats before traffic send

The change is in executor, please consider reversing the prefix order to:
`executor, trex: `

File: pkg/internal/checkup/executor/executor.go [hunk 2]
```diff
@@ -89,7 +89,7 @@ func (e Executor) Execute(ctx context.Context, vmiUnderTestName, trafficGenVMINa
 		return status.Results{}, fmt.Errorf("failed to Start to Trex Service on VMI \"%s/%s\": %w", e.namespace, trafficGenVMIName, err)
 	}
 
-	trexClient := trex.NewClient(e.vmiSerialClient, e.namespace, e.verbosePrintsEnabled)
+	trexClient := trex.NewClient(e.vmiSerialClient, e.namespace, e.trafficGeneratorPacketsPerSecond, e.testDuration, e.verbosePrintsEnabled)
```
Comment (orelmisan) [inline pkg/internal/checkup/executor/executor.go:87]: Could you please add a few words regarding the function of the above parameters in the commit message?

File: pkg/internal/checkup/trex/client.go [hunk 3]
```diff
@@ -35,20 +35,27 @@ import (
...
 
 const (
 	TrexStreamsPyPath = "/opt/tests"
 )
 
-func NewClient(vmiSerialClient vmiSerialConsoleClient, namespace string, verbosePrintsEnabled bool) Client {
+func NewClient(vmiSerialClient vmiSerialConsoleClient,
+	namespace, trafficGeneratorPacketsPerSecond string,
+	testDuration time.Duration,
+	verbosePrintsEnabled bool) Client {
```
Comment (orelmisan) [inline pkg/internal/checkup/trex/client.go:60]: We should address the growing size of params here (in a follow-up PR).

File: pkg/internal/checkup/executor/executor.go [hunk 4]
```diff
@@ -115,7 +115,13 @@ func (e Executor) Execute(ctx context.Context, vmiUnderTestName, trafficGenVMINa
 		return status.Results{}, fmt.Errorf("failed to run traffic from traffic generator VMI \"%s/%s\" side: %w",
 			e.namespace, trafficGenVMIName, err)
 	}
-	time.Sleep(e.testDuration)
+
+	var err error
+	trafficGeneratorMaxDropRate, err := trexClient.MonitorDropRates(ctx, trafficGenVMIName)
```
Comment (orelmisan) [inline pkg/internal/checkup/executor/executor.go:120]: Could you please elaborate why do we need to monitor the drop rates?

File: pkg/internal/checkup/executor/executor.go (hunk 1 above)
Comment (RamLavi) [reply pkg/internal/checkup/executor/executor.go:107]: DONE

File: pkg/internal/checkup/vmi.go [hunk 5]
```diff
@@ -152,7 +151,7 @@ func trafficGenBootCommands(configDiskSerial string) []string {
 		fmt.Sprintf("sudo cp %s/%s %s", configMountDirectory, trex.ExecutionScriptName, trex.BinDirectory),
 		fmt.Sprintf("sudo chmod 744 %s/%s", trex.BinDirectory, trex.ExecutionScriptName),
 		fmt.Sprintf("sudo cp %s/%s /etc", configMountDirectory, trex.CfgFileName),
-		fmt.Sprintf("sudo mkdir -p %s", testScriptsDirectory),
-		fmt.Sprintf("sudo cp %s/*.py %s", configMountDirectory, testScriptsDirectory),
+		fmt.Sprintf("sudo mkdir -p %s", trex.TrexStreamsPyPath),
```
[the remark this answers is not in the document]
Comment (RamLavi) [reply pkg/internal/checkup/vmi.go:154]: DONE

File: pkg/internal/checkup/executor/executor.go (hunk 2 above)
Comment (RamLavi) [reply pkg/internal/checkup/executor/executor.go:87]: Not much to say really. I wouldn't have moved them if the lint wouldn't have complained about it.

File: pkg/internal/checkup/trex/client.go [hunk 6]
```diff
@@ -52,6 +52,31 @@ func NewClient(vmiSerialClient vmiSerialConsoleClient, namespace string, verbose
 	}
 }
 
+func (t Client) MonitorDropRates(ctx context.Context, vmiName string, duration time.Duration) (float64, error) {
```
[the remark this answers is not in the document]
Comment (RamLavi) [reply pkg/internal/checkup/trex/client.go]: Now it does

File: pkg/internal/checkup/executor/executor.go (hunk 4 above)
Comment (RamLavi) [reply pkg/internal/checkup/executor/executor.go:120]: As we talked offline - there is no practical use of trafficGeneratorMaxDropRate, as it is only sampling and is not a measurable quantity we can use to determine 100% if there were any drops.
Having said that, it is much better than just waiting on a sleep.

**Comment by RamLavi** [comment]
```
$ TEST_NAMESPACE=dpdk-checkup-ns-1 NETWORK_ATTACHMENT_DEFINITION_NAME=dpdk-sriovnetwork-ns-1 make TEST_IMAGE=quay.io/ramlavi/kubevirt-dpdk-checkup:latest TRAFFIC_GEN_IMAGE_URL=quay.io/kiagnose/kubevirt-dpdk-checkup-traffic-gen:traffic-gen-in-vm VM_CONTAINER_DISK_IMAGE_URL=quay.io/kiagnose/kubevirt-dpdk-checkup-vm:traffic-gen-in-vm test/e2e
docker run --rm \
           --volume /home/ralavi/go/src/github.com/kiagnose/kubevirt-dpdk-checkup:/home/ralavi/go/src/github.com/kiagnose/kubevirt-dpdk-checkup:Z \
           --volume /home/ralavi/.kube:/root/.kube:Z,ro \
           --workdir /home/ralavi/go/src/github.com/kiagnose/kubevirt-dpdk-checkup \
           -e KUBECONFIG=/root/.kube/kubeconfig \
           -e TEST_IMAGE=quay.io/ramlavi/kubevirt-dpdk-checkup:latest \
           -e TEST_NAMESPACE=dpdk-checkup-ns-1 \
           -e NETWORK_ATTACHMENT_DEFINITION_NAME=dpdk-sriovnetwork-ns-1 \
           -e TRAFFIC_GEN_IMAGE_URL=quay.io/kiagnose/kubevirt-dpdk-checkup-traffic-gen:traffic-gen-in-vm \
           -e VM_CONTAINER_DISK_IMAGE_URL=quay.io/kiagnose/kubevirt-dpdk-checkup-vm:traffic-gen-in-vm \
           docker.io/library/golang:1.20.6 go test ./tests/... -test.v -test.timeout=1h -ginkgo.v -ginkgo.timeout=1h
=== RUN   TestKubevirtDpdkCheckup
Running Suite: KubevirtDpdkCheckup Suite - /home/ralavi/go/src/github.com/kiagnose/kubevirt-dpdk-checkup/tests
==============================================================================================================
Random Seed: 1690183942

Will run 1 of 1 specs
------------------------------
[BeforeSuite] 
/home/ralavi/go/src/github.com/kiagnose/kubevirt-dpdk-checkup/tests/test_suite_test.go:60
[BeforeSuite] PASSED [0.002 seconds]
------------------------------
Execute the checkup Job should complete successfully
/home/ralavi/go/src/github.com/kiagnose/kubevirt-dpdk-checkup/tests/checkup_test.go:79
  [FAILED] in [It] - /home/ralavi/go/src/github.com/kiagnose/kubevirt-dpdk-checkup/tests/checkup_test.go:89 @ 07/24/23 07:36:28.451
• [FAILED] [247.105 seconds]
Execute the checkup Job [It] should complete successfully
/home/ralavi/go/src/github.com/kiagnose/kubevirt-dpdk-checkup/tests/checkup_test.go:79

  [FAILED] checkup failed: {
  	"spec.param.networkAttachmentDefinitionName": "dpdk-sriovnetwork-ns-1",
  	"spec.param.testDuration": "1m",
  	"spec.param.trafficGeneratorImage": "quay.io/kiagnose/kubevirt-dpdk-checkup-traffic-gen:traffic-gen-in-vm",
  	"spec.param.verbose": "true",
  	"spec.param.vmContainerDiskImage": "quay.io/kiagnose/kubevirt-dpdk-checkup-vm:traffic-gen-in-vm",
  	"spec.timeout": "10m",
  	"status.completionTimestamp": "2023-07-24T07:36:21Z",
  	"status.failureReason": "detected packets dropped on the DPDK VM's side: RX: 19601014; TX: 0",
  	"status.result.DPDKRxPacketDrops": "19601014",
  	"status.result.DPDKRxTestPackets": "505065040",
  	"status.result.DPDKTxPacketDrops": "0",
  	"status.result.DPDKVMNode": "cnv-qe-infra-23.cnvqe2.lab.eng.rdu2.redhat.com",
  	"status.result.trafficGeneratorInErrorPackets": "0",
  	"status.result.trafficGeneratorNode": "cnv-qe-infra-23.cnvqe2.lab.eng.rdu2.redhat.com",
  	"status.result.trafficGeneratorOutputErrorPackets": "0",
  	"status.result.trafficGeneratorTxPackets": "0",
  	"status.startTimestamp": "2023-07-24T07:32:29Z",
  	"status.succeeded": "false"
  }
  In [It] at: /home/ralavi/go/src/github.com/kiagnose/kubevirt-dpdk-checkup/tests/checkup_test.go:89 @ 07/24/23 07:36:28.451
------------------------------

Summarizing 1 Failure:
  [FAIL] Execute the checkup Job [It] should complete successfully
  /home/ralavi/go/src/github.com/kiagnose/kubevirt-dpdk-checkup/tests/checkup_test.go:89

Ran 1 of 1 Specs in 247.108 seconds
FAIL! -- 0 Passed | 1 Failed | 0 Pending | 0 Skipped
--- FAIL: TestKubevirtDpdkCheckup (247.11s)
FAIL
FAIL	github.com/kiagnose/kubevirt-dpdk-checkup/tests	247.130s
FAIL
```

File: pkg/internal/checkup/trex/client.go [hunk 7]
```diff
@@ -130,7 +130,7 @@ func (t Client) getTrexServiceJournalctl(vmiName string) (string, error) {
 }
 
 func (t Client) runTrexConsoleCmd(vmiName, command string) (string, error) {
-	shellCommand := fmt.Sprintf("/bin/sh -c \"cd /opt/trex && echo %q | ./trex-console -q\"", command)
+	shellCommand := fmt.Sprintf("cd /opt/trex && echo %q | ./trex-console -q", command)
```
Comment (RamLavi) [inline pkg/internal/checkup/trex/client.go:210]: use path const

Comment (orelmisan) [inline pkg/internal/checkup/trex/client.go:210]: Commit message: The command was executed in a sub-shell [1].
The advantage of it was that it did not change the current shell.
The disadvantage is that it is a bit slower.

[1] https://tldp.org/LDP/abs/html/subshells.html#:~:text=A%20subshell%20is%20a%20separate,process%20a%20list%20of%20commands.

Comment (RamLavi) [reply pkg/internal/checkup/trex/client.go:210]: DONE

File: pkg/internal/checkup/trex/client.go [hunk 8]
```diff
@@ -151,6 +151,12 @@ func (t Client) ClearStats(vmiName string) (string, error) {
 	return t.runTrexConsoleCmd(vmiName, "clear")
 }
 
+func (t Client) StartTraffic(vmiName string, port PortIdx) (string, error) {
+	testDurationSeconds := int(t.testDuration.Seconds())
```
Comment (orelmisan) [inline pkg/internal/checkup/trex/client.go:204]: This temp variable is redundant IMO.

File: pkg/internal/checkup/trex/client.go [hunk 9]
```diff
@@ -151,6 +151,12 @@ func (t Client) ClearStats(vmiName string) (string, error) {
...
+	return t.runTrexConsoleCmd(vmiName, fmt.Sprintf("start -f %s/testpmd.py -m %spps -p %d -d %d",
```
Comment (orelmisan) [inline pkg/internal/checkup/trex/client.go:205]: Please consider building the command in stages using a string builder.

File: pkg/internal/checkup/trex/client.go (hunk 7 above)
Comment (RamLavi) [reply pkg/internal/checkup/trex/client.go:210]: DONE

File: pkg/internal/checkup/trex/client.go (hunk 8 above)
Comment (RamLavi) [reply pkg/internal/checkup/trex/client.go:204]: changed to builder.

File: pkg/internal/checkup/trex/client.go (hunk 9 above)
Comment (RamLavi) [reply pkg/internal/checkup/trex/client.go:205]: DONE

**Comment by RamLavi** [comment]
[Change](https://github.com/kiagnose/kubevirt-dpdk-checkup/compare/3e4f8e54b9cfb9ac6f525df8d092508a14b1e0ae..43431bca2ffb98c018b12fcd580e11dc60aac2c3): review fixes

**Review by orelmisan — APPROVED** [review]

**Pull request merged**